### ISSUE TITLE
How to set loss function and metrics in deepmedic? 

### ISSUE THREAD

**Issue by guyucowboy** [Kamnitsask/deepmedic#84]
Hi, kamnitsask:
     Great work! Thank you for sharing your code. 
     I have a question to ask you. How can I set loss function and metrics in deepmedic or in theano ?
     For example,in keras, the lose function could be set as mean_squared_error , dice_coef_loss, etc. and metrics could be set as [dice_coef] or  metrics=[accuracy] when compiling the model.
     How to change the loss function and metric in deepmedic? I want to change the default setting to  dice_coef_loss and dice_coef.
Thanks
Gu Yu

**Comment by Kamnitsask** [comment]
Hi and sorry for the very late reply, I ve been quite overwhelmed lately, hopefully I ll manage to support this again more often.

Keras is a higher layer over theano, so it provides quite some functionality in a modular way. In this project, built straight with theano, you ll have to write the new loss function yourself and call it.

The loss function is called in ./deepmedic/cnn3d.py in the compileTrainFunction() function, where the negativeLogLikelihood is called. You ll have to write a similar cost function for dice and call it (I have an implementation of dice lying somewhere around, I may as well push it soon). 

Regarding metrics: Situation is similar and more involved than Keras. You need to return the values you want from the train/validation function back to the cpu, and log them. Then if you want to plot them, update the plotting script ./plotTrainingProgress.py . In the ./deepmedic/cnn3d.py -> compileTrainFunction() and compileValidationFunction(), check how the cost is being returned, or the true positives / true negatives are returned by getRpRnTpTnForTrain0OrVal1(...), and returned to the function that calls a training iteration  ./deepmedic/trainValidateTestVisualiseParallel.py-> doTrainOrValidationOnBatchesAndReturnMeanAccuraciesOfSubepoch()

Sorry again for the late reply.
Kostas

**Comment by guyucowboy** [comment]
Thanks for your reply!
I maybe trouble you again when I have some other questions.
Thanks again for your help!
Gu Yu
